### ISSUE TITLE
.dockerignore says it ignores *, but doesn't.

### ISSUE THREAD

**Issue by kojiromike** [docker/docker#13417]
```
$ docker build .
Sending build context to Docker daemon 406.7 MB
```

But I didn't need any of the local files, so I just added `*` to .dockerignore. Then…

```
$ docker build .
ERRO[0000] Can't exclude whole path, excluding pattern: * 
Sending build context to Docker daemon 4.608 kB
```

But 4.608 kB << 406.7 MB, so So clearly it did _not_ _not_ ignore my files.

**Comment by duglin** [comment]
@kojiromike what version of Docker are you using?
Also note that it says:
```
ERRO[0000] Can't exclude whole path, excluding pattern: * 
```
meaning it ignore that line in the Dockerfile.
What's interesting is that in the latest version of Docker I don't see that error message so the latest version might work the way you expect - sort of.  In my test it excluded all files and dir not just files.  :-)
I'm not sure this is correct, will need to think about this more.

**Comment by kojiromike** [comment]
```
$ docker --version
Docker version 1.6.2, build 7c8fca2
```

**Comment by kojiromike** [comment]
On a tangential note, I think `docker-compose` actually obeys `.dockerignore` more literally and even ignores the `Dockerfile` itself, causing a critical failure:

```
$ compose build
dbdata uses an image, skipping
db uses an image, skipping
Building shell...
Cannot locate specified Dockerfile: Dockerfile
```

(I don't regard this as a bug in Compose, just an interesting anecdote about how `docker` and `docker-compose` build differently.)

**Comment by duglin** [comment]
@buddhamagnet do you remember why on this line: https://github.com/docker/docker/blob/master/pkg/fileutils/fileutils.go#L66  we don't throw an error?  In the old code we did but I can't remember if we took out the error on purpose or whether during all of the re-organizing we just forgot it?

**Comment by duglin** [comment]
@kojiromike I've never looked at the compose code but I know in Docker we go out of our way to special case the Dockerfile and .dockerignore. They're always sent to the daemon (even if mentioned in the .dockerignore) but then they're removed from the build context immediately after the Dockerfile is parsed - but before any Dockerfile commands are run. Which is why a COPY won't see them.

**Comment by duglin** [comment]
Hmmm, you're using a new enough version. Not sure why we're seeing different results.
Do you have anything other than "*" in your .dockerignore?

**Comment by kojiromike** [comment]
Yes. It was

```
Dockerfile
docker-compose.yml
*
```

(I was trying to get `docker build` not to miss cache when all I had changed was to add a new command at the end of the Dockerfile.)

**Comment by duglin** [comment]
OK, running with v1.6.2 I see the same thing you do. But running on master I don't.
Between 1.6.2 and now something broke because I should see the error.  hmmmm

**Comment by kojiromike** [comment]
`git bisect` ftw? :)

**Comment by duglin** [comment]
ok - I just confirmed that @buddhamagnet and I removed the error msg during a previous PR - I'm hoping just by mistake. However, fixing that will just get us to where you are.
I don't think you can exclude just the local files via .dockerignore.  "*" will grab everything, dirs too.

**Comment by kojiromike** [comment]
but how do you explain the size difference?

**Comment by duglin** [comment]
ok I see what's going on.  We print the Error but then keep going.  But, I think because of the logic in there it actually did exclude everything - the logic is kind of backwards.  I think we need to fix this so that it prints the error and stops.

**Comment by duglin** [comment]
But let's wait for @buddhamagnet to speak up to see if he can remember what we were thinking :-)

**Comment by buddhamagnet** [comment]
@duglin @kojiromike sorry guys only just seen this! @duglin yep I think you're right.

**Comment by buddhamagnet** [comment]
@duglin there were so many switches in direction, consolidations and reversions in that flow think we just missed that and looks like we had a hole in the tests too!

**Comment by duglin** [comment]
So.... I have a fix for this. Meaning I detect when someone tries to exclude the entire context (via a "*") and will generate an error.  It, however, does allow for someone to do this in their .dockerignore:
```
*
!foo
```
meaning exclude everything except foo.  
But, before I push the PR I started to wonder if this was the correct fix.   Thoughts...
- should we bother to try to detect this at all?  While I'm pretty sure the code works, I'm wondering if we should just let the build fail later on due to missing files.  What's worrying me is that there may be other cases (like the 'everything except...' case above) that we're going to have to continually update the code to support just so we can fail fast.
- if someone wants to do something stupid, I'm inclined to let them as long as they do eventually get an error
- I'm not sure we should ban this case in the first place.  If all someone has just a Dockerfile, why shouldn't they be allowed to .dockerignore everything?  We will go out of our way to send the Dockerfile anyway, so the "*" in the .dockerignore won't change that.

Net - I'm thinking we should remove these types of checks.  But what do you guys think?

**Comment by kojiromike** [comment]
I think docker should support a way to prevent sending the build context. (I also think the fact that the user can do this providing the Dockerfile as standard input is tribal knowledge and shouldn't be the _only_ way to do it.)

So I think that `.dockerignore` should allow you to ignore `*`, and that should mean everything but the Dockerfile itself gets ignored.

**Comment by cpuguy83** [comment]
If you don't need any local files why are you passing in `.`?
This tells the CLI to package up `.` and send it as the context.
If you want just the Dockerfile, do `docker build -< Dockerfile`

**Comment by kojiromike** [comment]
@cpuguy83 Thanks, but I'm not looking for just a way to prevent uploading the build context. I'm looking for a way to _define_ that this build has no (need for a) build context. The former is implementation. The latter is interface. For the former, each user will need to know the `docker` command invocation individually. For the `*` in the .dockerignore file, it will "just work".

**Comment by jfrazelle** [comment]
closed by  #13611
